// maplibre-gl-js-amplify/lib/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmplifyGeofenceControl = exports.drawGeofences = exports.createAmplifyGeocoder = exports.createDefaultIcon = exports.createMap = exports.AmplifyGeocoderAPI = exports.drawPoints = exports.AmplifyMapLibreRequest = void 0;
const AmplifyMapLibreRequest_1 = __importStar(require("./AmplifyMapLibreRequest"));
exports.AmplifyMapLibreRequest = AmplifyMapLibreRequest_1.default;
Object.defineProperty(exports, "createMap", { enumerable: true, get: function () { return AmplifyMapLibreRequest_1.createMap; } });
const drawPoints_1 = require("./drawPoints");
Object.defineProperty(exports, "drawPoints", { enumerable: true, get: function () { return drawPoints_1.drawPoints; } });
const AmplifyMapLibreGeocoder_1 = require("./AmplifyMapLibreGeocoder");
Object.defineProperty(exports, "AmplifyGeocoderAPI", { enumerable: true, get: function () { return AmplifyMapLibreGeocoder_1.AmplifyGeocoderAPI; } });
Object.defineProperty(exports, "createAmplifyGeocoder", { enumerable: true, get: function () { return AmplifyMapLibreGeocoder_1.createAmplifyGeocoder; } });
const createDefaultIcon_1 = require("./createDefaultIcon");
Object.defineProperty(exports, "createDefaultIcon", { enumerable: true, get: function () { return createDefaultIcon_1.createDefaultIcon; } });
const drawGeofences_1 = require("./drawGeofences");
Object.defineProperty(exports, "drawGeofences", { enumerable: true, get: function () { return drawGeofences_1.drawGeofences; } });
const index_1 = require("./AmplifyGeofenceControl/index");
Object.defineProperty(exports, "AmplifyGeofenceControl", { enumerable: true, get: function () { return index_1.AmplifyGeofenceControl; } });
